Add initial ImageDataArray as replacement type for ImageData::m_data

squelart opened the pull request:
#### faf054788cd86c2302c77d2d5368e342924f982a
<pre>
Add initial ImageDataArray as replacement type for ImageData::m_data
<a href="https://bugs.webkit.org/show_bug.cgi?id=283939">https://bugs.webkit.org/show_bug.cgi?id=283939</a>
<a href="https://rdar.apple.com/problem/140813668">rdar://problem/140813668</a>

Reviewed by Said Abou-Hallawa.

Currently `ImageData::m_data` is a `Ref&lt;JSC::Uint8ClampedArray&gt;` (corresponding
to ImageData.idl&apos;s `attribute ImageDataArray data`), so it can only deal with
byte-sized color components like RGBA.
To allow float16 support (the idl data type will change to something like
`typedef (Uint8ClampedArray or Float16Array) ImageDataArray`), `ImageData::m_data`
should be able to hold other typed-array types, like `JSC::Float16Array`.
As a first step, we&apos;ll move the `Ref&lt;JSC::Uint8ClampedArray&gt;` to a separate class,
and that class will eventually handle more types.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
(WebCore::ImageData::ImageData):
(WebCore::ImageData::clone const):
* Source/WebCore/html/ImageData.h:
* Source/WebCore/html/ImageDataArray.cpp: Added.
* Source/WebCore/html/ImageDataArray.h: Added.
(WebCore::ImageDataArray::ImageDataArray):
(WebCore::ImageDataArray::length const):
(WebCore::ImageDataArray::get const):
(WebCore::ImageDataArray::data const):

Canonical link: <a href="https://commits.webkit.org/287312@main">https://commits.webkit.org/287312@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94b7c6acf2da2ad32a25dc9f980d11744e40e643

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79006 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32383 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83646 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30221 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81139 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6312 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61838 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19755 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82073 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51875 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71977 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42142 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49226 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26089 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28587 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70340 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26502 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85032 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6331 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4373 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70071 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6495 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67890 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69321 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17298 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13367 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12125 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6283 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12133 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6244 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9695 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8035 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/html/ImageDataArray.h:
```diff
@@ -0,0 +1,56 @@
+
+    // FIXME: Remove these temporary functions.
+    Uint8ClampedArray& get() const { return m_bufferView.get(); }
+    auto data() const { return m_bufferView->data(); }
```
How is `ImageData` supposed to deal with its method 
```
Uint8ClampedArray& data() const { return m_data.get(); }
```
when the underlying data of this class is `JSC::Float16Array`?

squelart replied:
As it says just above: "// FIXME: Remove these temporary functions." 😉
The goal with this patch was only to introduce this intermediary `ImageDataArray` type that *currently* only deals with `Uint8ClampedArray`. The next patch will add float16 support (WIP: https://github.com/WebKit/WebKit/commit/12134772c758200331c6b268147b91a20afee1b0 ).

squelart replied:
Basically (spoiler alert!), `ImageDataArray` will store`Ref<ArrayBufferView>` (common base class), and instead of this `data()` there will be a `visit(F&&)` template function that will call a given lambda with `Uint8ClampedArray` or `Float16Array`.

shallawa replied:
How will the caller know whether `ImageData` stores bytes or float16? I think it will help reviewing these incremental patch if you can put somewhere how `ImageData` will look eventually.

shallawa replied:
Sorry I just opened your WIP: https://github.com/WebKit/WebKit/commit/12134772c758200331c6b268147b91a20afee1b0 now. I saw the answer to my question.

shallawa commented on Source/WebCore/Headers.cmake:
```diff
@@ -1335,6 +1335,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     html/ImageData.h
+    html/ImageDataArray.h
     html/ImageDataSettings.h
     html/ImageDataStorageFormat.h
```
None of these headers is used outside WebCore. Are we sure we need to export them to other modules?

squelart committed:
Add initial ImageDataArray as replacement type for ImageData::m_data
https://bugs.webkit.org/show_bug.cgi?id=283939
rdar://problem/140813668

Reviewed by Said Abou-Hallawa.

Currently `ImageData::m_data` is a `Ref<JSC::Uint8ClampedArray>` (corresponding
to ImageData.idl's `attribute ImageDataArray data`), so it can only deal with
byte-sized color components like RGBA.
To allow float16 support (the idl data type will change to something like
`typedef (Uint8ClampedArray or Float16Array) ImageDataArray`), `ImageData::m_data`
should be able to hold other typed-array types, like `JSC::Float16Array`.
As a first step, we'll move the `Ref<JSC::Uint8ClampedArray>` to a separate class,
and that class will eventually handle more types.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
(WebCore::ImageData::ImageData):
(WebCore::ImageData::clone const):
* Source/WebCore/html/ImageData.h:
* Source/WebCore/html/ImageDataArray.cpp: Added.
* Source/WebCore/html/ImageDataArray.h: Added.
(WebCore::ImageDataArray::ImageDataArray):
(WebCore::ImageDataArray::length const):
(WebCore::ImageDataArray::get const):
(WebCore::ImageDataArray::data const):

Canonical link: https://commits.webkit.org/287312@main
